File: src/arrow_analytics.py
"""
Arrow shaft analytics: quality grading, group statistics, outlier detection,
set optimization, and similarity matching.

All functions are pure — no database or web dependencies.
"""

from __future__ import annotations


# ---------------------------------------------------------------------------
# Grade thresholds  (configurable defaults)
# ---------------------------------------------------------------------------

GRADE_THRESHOLDS: list[tuple[str, float, float]] = [
    #            (grade,         max_weight_dev_gr, max_straightness)
    ("Premium",      0.5, 0.001),
    ("Competition",  1.0, 0.003),
    ("Amateur",      2.0, 0.006),
]

FALLBACK_GRADE = "Recreational"


# ---------------------------------------------------------------------------
# Shaft quality grading
# ---------------------------------------------------------------------------


def grade_shaft(
    shaft_weight: float,
    mean_weight: float,
    straightness: float | None,
) -> str:
    """Grade a single shaft based on weight deviation from set mean and straightness.

    Rules (evaluated in order, first match wins):
    - Premium:     weight dev < 0.5 gr **and** straightness < 0.001
    - Competition: weight dev < 1.0 gr **and** straightness < 0.003
    - Amateur:     weight dev < 2.0 gr **and** straightness < 0.006
    - Recreational: everything else

    When *straightness* is ``None`` the shaft is graded on weight alone.
    """
    weight_dev = abs(shaft_weight - mean_weight)

    for grade, max_dev, max_straight in GRADE_THRESHOLDS:
        weight_ok = weight_dev < max_dev
        straight_ok = straightness is None or straightness < max_straight
        if weight_ok and straight_ok:
            return grade

    return FALLBACK_GRADE


# ---------------------------------------------------------------------------
# Group statistics
# ---------------------------------------------------------------------------


def _metric_stats(values: list[float]) -> dict | None:
    """Return {mean, std, range, cv_pct} for a list of floats, or None if empty."""
    if not values:
        return None
    n = len(values)
    mean = sum(values) / n
    if n < 2:
        return {"mean": round(mean, 3), "std": 0.0, "range": 0.0, "cv_pct": 0.0}
    variance = sum((v - mean) ** 2 for v in values) / (n - 1)
    std = variance ** 0.5
    val_range = max(values) - min(values)
    cv_pct = (std / mean * 100) if mean != 0 else 0.0
    return {
        "mean": round(mean, 3),
        "std": round(std, 3),
        "range": round(val_range, 3),
        "cv_pct": round(cv_pct, 2),
    }


def compute_group_stats(shafts: list[dict]) -> dict:
    """Compute group-level statistics for a list of shaft dicts.

    Each dict should have keys: ``measured_weight_gr``, ``measured_spine_astm``,
    ``straightness`` — any may be ``None`` and will be skipped.

    Returns::

        {
            "weight":       MetricStats | None,
            "spine":        MetricStats | None,
            "straightness": MetricStats | None,
            "shaft_count":  int,
        }
    """
    weights = [s["measured_weight_gr"] for s in shafts if s.get("measured_weight_gr") is not None]
    spines = [s["measured_spine_astm"] for s in shafts if s.get("measured_spine_astm") is not None]
    straights = [s["straightness"] for s in shafts if s.get("straightness") is not None]

    return {
        "weight": _metric_stats(weights),
        "spine": _metric_stats(spines),
        "straightness": _metric_stats(straights),
        "shaft_count": len(shafts),
    }


# ---------------------------------------------------------------------------
# Outlier detection  (Phase 2 — placeholder)
# ---------------------------------------------------------------------------


def detect_outliers(
    shafts: list[dict],
    threshold: float = 2.0,
) -> list[dict]:
    """Z-score outlier detection across weight, spine, and straightness.

    Returns a list of ``{arrow_number, feature, z_score, reason}`` entries.
    Requires >= 3 shafts with data for a given feature; otherwise skips that feature.
    """
    if len(shafts) < 3:
        return []

    results: list[dict] = []
    features = [
        ("weight", "measured_weight_gr", "gr"),
        ("spine", "measured_spine_astm", ""),
        ("straightness", "straightness", ""),
    ]

    for feature_name, key, unit in features:
        values = [(s, s.get(key)) for s in shafts if s.get(key) is not None]
        if len(values) < 3:
            continue

        vals = [v for _, v in values]
        n = len(vals)
        mean = sum(vals) / n
        std = (sum((v - mean) ** 2 for v in vals) / (n - 1)) ** 0.5
        if std == 0:
            continue

        for shaft, val in values:
            z = abs(val - mean) / std
            if z >= threshold:
                direction = "above" if val > mean else "below"
                suffix = f" {unit}" if unit else ""
                reason = (
                    f"{feature_name.capitalize()} {z:.1f}σ {direction} mean "
                    f"({val}{suffix} vs {mean:.1f}{suffix} mean)"
                )
                results.append({
                    "arrow_number": shaft.get("arrow_number", 0),
                    "feature": feature_name,
                    "value": val,
                    "z_score": round(z, 2),
                    "reason": reason,
                })

    return results


# ---------------------------------------------------------------------------
# Set consistency scoring & best-set finder  (Phase 5)
# ---------------------------------------------------------------------------

DEFAULT_WEIGHTS = {"weight": 0.4, "spine": 0.35, "straightness": 0.25}


def score_set_consistency(
    shafts: list[dict],
    weights: dict | None = None,
) -> float:
    """Score a set of shafts by consistency. Lower = more consistent.

    Normalizes each feature's standard deviation by the feature range across
    the *provided* shafts (avoids requiring global min/max from the full population).
    Score = weighted sum of normalised stds.

    Returns 0.0 for sets with fewer than 2 shafts.
    """
    if len(shafts) < 2:
        return 0.0

    w = weights or DEFAULT_WEIGHTS
    features = [
        ("weight", "measured_weight_gr"),
        ("spine", "measured_spine_astm"),
        ("straightness", "straightness"),
    ]

    score = 0.0
    for feat_name, key in features:
        vals = [s[key] for s in shafts if s.get(key) is not None]
        if len(vals) < 2:
            continue
        mean = sum(vals) / len(vals)
        std = (sum((v - mean) ** 2 for v in vals) / len(vals)) ** 0.5
        # Normalise by mean (CV) to make features comparable
        if mean != 0:
            norm_std = std / abs(mean)
        else:
            norm_std = std
        score += w.get(feat_name, 0) * norm_std

    return round(score, 6)


def find_best_sets(
    shafts: list[dict],
    set_size: int = 6,
    top_n: int = 5,
    weights: dict | None = None,
    max_combinations: int = 200_000,
) -> list[dict]:
    """Find the top N most consistent sets of ``set_size`` arrows.

    Uses ``itertools.combinations`` for small N; switches to random sampling
    when the number of combinations exceeds ``max_combinations``.

    Returns a list of dicts ranked by consistency (best first).
    """
    import itertools
    import math
    import random

    n = len(shafts)
    if n < set_size:
        return []

    total_combos = math.comb(n, set_size)

    if total_combos <= max_combinations:
        # Exhaustive search
        candidates = list(itertools.combinations(range(n), set_size))
    else:
        # Random sampling — use index tuples to avoid duplicates
        seen: set[tuple[int, ...]] = set()
        while len(seen) < max_combinations:
            combo = tuple(sorted(random.sample(range(n), set_size)))
            seen.add(combo)
        candidates = list(seen)

    scored: list[tuple[float, tuple[int, ...]]] = []
    for combo in candidates:
        subset = [shafts[i] for i in combo]
        s = score_set_consistency(subset, weights)
        scored.append((s, combo))

    scored.sort(key=lambda x: x[0])

    results: list[dict] = []
    w = weights or DEFAULT_WEIGHTS
    for rank, (consistency, combo) in enumerate(scored[:top_n], start=1):
        subset = [shafts[i] for i in combo]
        arrow_numbers = [s.get("arrow_number", 0) for s in subset]

        # Per-feature stds for the result
        weight_vals = [s["measured_weight_gr"] for s in subset if s.get("measured_weight_gr") is not None]
        spine_vals = [s["measured_spine_astm"] for s in subset if s.get("measured_spine_astm") is not None]
        straight_vals = [s["straightness"] for s in subset if s.get("straightness") is not None]

        def _std(vals: list[float]) -> float:
            if len(vals) < 2:
                return 0.0
            m = sum(vals) / len(vals)
            return round((sum((v - m) ** 2 for v in vals) / len(vals)) ** 0.5, 4)

        results.append({
            "rank": rank,
            "arrow_numbers": sorted(arrow_numbers),
            "consistency_score": consistency,
            "weight_std_gr": _std(weight_vals),
            "spine_std": _std(spine_vals),
            "straightness_std": _std(straight_vals),
        })

    return results


def find_similar_arrows(
    reference_arrows: list[dict],
    candidates: list[dict],
    top_n: int = 5,
) -> list[dict]:
    """Find arrows most similar to a reference set's centroid.

    1. Compute centroid of reference set in [weight, spine, straightness] space
    2. Normalize all features to [0, 1] using min/max from combined pool
    3. Rank candidates by Euclidean distance from centroid

    Returns a list of dicts sorted by similarity (best first), each with:
        - arrow_number, similarity_score (0-100), weight_diff_gr,
          spine_diff, straightness_diff
    """
    if not reference_arrows or not candidates:
        return []

    features = ("measured_weight_gr", "measured_spine_astm", "straightness")

    # Compute centroid of reference set
    centroid: dict[str, float] = {}
    for feat in features:
        vals = [a[feat] for a in reference_arrows if a.get(feat) is not None]
        centroid[feat] = sum(vals) / len(vals) if vals else 0.0

    # Collect all values (reference + candidates) for normalisation
    all_arrows = reference_arrows + candidates
    ranges: dict[str, tuple[float, float]] = {}
    for feat in features:
        vals = [a[feat] for a in all_arrows if a.get(feat) is not None]
        if vals:
            lo, hi = min(vals), max(vals)
            ranges[feat] = (lo, hi)
        else:
            ranges[feat] = (0.0, 1.0)

    def _normalize(value: float, feat: str) -> float:
        lo, hi = ranges[feat]
        return (value - lo) / (hi - lo) if hi != lo else 0.5

    centroid_norm = {f: _normalize(centroid[f], f) for f in features}

    # Rank candidates
    scored: list[tuple[float, dict]] = []
    for arrow in candidates:
        vals_norm = {}
        raw_vals: dict[str, float] = {}
        skip = False
        for feat in features:
            v = arrow.get(feat)
            if v is None:
                skip = True
                break
            vals_norm[feat] = _normalize(v, feat)
            raw_vals[feat] = v
        if skip:
            continue

        dist = sum((vals_norm[f] - centroid_norm[f]) ** 2 for f in features) ** 0.5
        # Convert distance to 0-100 similarity.  max possible dist = sqrt(3) ~ 1.732
        max_dist = len(features) ** 0.5
        similarity = max(0.0, (1.0 - dist / max_dist) * 100.0)

        scored.append((similarity, {
            "arrow_number": arrow.get("arrow_number", 0),
            "similarity_score": round(similarity, 1),
            "weight_diff_gr": round(raw_vals["measured_weight_gr"] - centroid["measured_weight_gr"], 2),
            "spine_diff": round(raw_vals["measured_spine_astm"] - centroid["measured_spine_astm"], 2),
            "straightness_diff": round(raw_vals["straightness"] - centroid["straightness"], 4),
        }))

    scored.sort(key=lambda x: x[0], reverse=True)
    return [entry for _, entry in scored[:top_n]]

File: frontend/src/domain/arrowAnalytics.ts
/**
 * Arrow shaft analytics: quality grading, group statistics, outlier detection,
 * set optimization, and similarity matching.
 * Ported from src/arrow_analytics.py — pure math, no external dependencies.
 */

// ---------------------------------------------------------------------------
// Grade thresholds
// ---------------------------------------------------------------------------

const GRADE_THRESHOLDS: Array<[string, number, number]> = [
  // [grade, max_weight_dev_gr, max_straightness]
  ["Premium", 0.5, 0.001],
  ["Competition", 1.0, 0.003],
  ["Amateur", 2.0, 0.006],
];

const FALLBACK_GRADE = "Recreational";

// ---------------------------------------------------------------------------
// Shaft quality grading
// ---------------------------------------------------------------------------

/**
 * Grade a single shaft based on weight deviation from set mean and straightness.
 */
export function gradeShaft(
  shaftWeight: number,
  meanWeight: number,
  straightness: number | null,
): string {
  const weightDev = Math.abs(shaftWeight - meanWeight);

  for (const [grade, maxDev, maxStraight] of GRADE_THRESHOLDS) {
    const weightOk = weightDev < maxDev;
    const straightOk = straightness === null || straightness < maxStraight;
    if (weightOk && straightOk) return grade;
  }

  return FALLBACK_GRADE;
}

// ---------------------------------------------------------------------------
// Group statistics
// ---------------------------------------------------------------------------

export interface MetricStats {
  mean: number;
  std: number;
  range: number;
  cv_pct: number;
}

function metricStats(values: number[]): MetricStats | null {
  if (values.length === 0) return null;
  const n = values.length;
  const mean = values.reduce((a, b) => a + b, 0) / n;
  if (n < 2) return { mean: Math.round(mean * 1000) / 1000, std: 0.0, range: 0.0, cv_pct: 0.0 };

  const variance = values.reduce((acc, v) => acc + (v - mean) ** 2, 0) / (n - 1);
  const std = Math.sqrt(variance);
  const valRange = Math.max(...values) - Math.min(...values);
  const cvPct = mean !== 0 ? (std / mean) * 100 : 0.0;

  return {
    mean: Math.round(mean * 1000) / 1000,
    std: Math.round(std * 1000) / 1000,
    range: Math.round(valRange * 1000) / 1000,
    cv_pct: Math.round(cvPct * 100) / 100,
  };
}

export interface ShaftData {
  arrow_number?: number;
  measured_weight_gr?: number | null;
  measured_spine_astm?: number | null;
  straightness?: number | null;
}

export interface GroupStats {
  weight: MetricStats | null;
  spine: MetricStats | null;
  straightness: MetricStats | null;
  shaft_count: number;
}

/**
 * Compute group-level statistics for a list of shaft data.
 */
export function computeGroupStats(shafts: ShaftData[]): GroupStats {
  const weights = shafts.filter((s) => s.measured_weight_gr != null).map((s) => s.measured_weight_gr!);
  const spines = shafts.filter((s) => s.measured_spine_astm != null).map((s) => s.measured_spine_astm!);
  const straights = shafts.filter((s) => s.straightness != null).map((s) => s.straightness!);

  return {
    weight: metricStats(weights),
    spine: metricStats(spines),
    straightness: metricStats(straights),
    shaft_count: shafts.length,
  };
}

// ---------------------------------------------------------------------------
// Outlier detection
// ---------------------------------------------------------------------------

export interface OutlierResult {
  arrow_number: number;
  feature: string;
  value: number;
  z_score: number;
  reason: string;
}

/**
 * Z-score outlier detection across weight, spine, and straightness.
 * Requires >= 3 shafts with data for a given feature.
 */
export function detectOutliers(shafts: ShaftData[], threshold = 2.0): OutlierResult[] {
  if (shafts.length < 3) return [];

  const results: OutlierResult[] = [];
  const features: Array<[string, keyof ShaftData, string]> = [
    ["weight", "measured_weight_gr", "gr"],
    ["spine", "measured_spine_astm", ""],
    ["straightness", "straightness", ""],
  ];

  for (const [featureName, key, unit] of features) {
    const values: Array<[ShaftData, number]> = [];
    for (const s of shafts) {
      const v = s[key];
      if (v != null && typeof v === "number") values.push([s, v]);
    }
    if (values.length < 3) continue;

    const vals = values.map(([, v]) => v);
    const n = vals.length;
    const mean = vals.reduce((a, b) => a + b, 0) / n;
    const std = Math.sqrt(vals.reduce((acc, v) => acc + (v - mean) ** 2, 0) / (n - 1));
    if (std === 0) continue;

    for (const [shaft, val] of values) {
      const z = Math.abs(val - mean) / std;
      if (z >= threshold) {
        const direction = val > mean ? "above" : "below";
        const suffix = unit ? ` ${unit}` : "";
        const reason = `${featureName.charAt(0).toUpperCase() + featureName.slice(1)} ${z.toFixed(1)}σ ${direction} mean (${val}${suffix} vs ${mean.toFixed(1)}${suffix} mean)`;
        results.push({
          arrow_number: shaft.arrow_number ?? 0,
          feature: featureName,
          value: val,
          z_score: Math.round(z * 100) / 100,
          reason,
        });
      }
    }
  }

  return results;
}

// ---------------------------------------------------------------------------
// Set consistency scoring & best-set finder
// ---------------------------------------------------------------------------

const DEFAULT_WEIGHTS = { weight: 0.4, spine: 0.35, straightness: 0.25 };

/**
 * Score a set of shafts by consistency. Lower = more consistent.
 */
export function scoreSetConsistency(
  shafts: ShaftData[],
  weights?: Partial<typeof DEFAULT_WEIGHTS>,
): number {
  if (shafts.length < 2) return 0.0;

  const w = { ...DEFAULT_WEIGHTS, ...weights };
  const features: Array<[string, keyof ShaftData]> = [
    ["weight", "measured_weight_gr"],
    ["spine", "measured_spine_astm"],
    ["straightness", "straightness"],
  ];

  let score = 0.0;
  for (const [featName, key] of features) {
    const vals = shafts.filter((s) => s[key] != null).map((s) => s[key] as number);
    if (vals.length < 2) continue;

    const mean = vals.reduce((a, b) => a + b, 0) / vals.length;
    const std = Math.sqrt(vals.reduce((acc, v) => acc + (v - mean) ** 2, 0) / vals.length);
    const normStd = mean !== 0 ? std / Math.abs(mean) : std;
    score += (w[featName as keyof typeof w] ?? 0) * normStd;
  }

  return Math.round(score * 1000000) / 1000000;
}

function stdPop(vals: number[]): number {
  if (vals.length < 2) return 0.0;
  const m = vals.reduce((a, b) => a + b, 0) / vals.length;
  return Math.round(Math.sqrt(vals.reduce((acc, v) => acc + (v - m) ** 2, 0) / vals.length) * 10000) / 10000;
}

export interface BestSetResult {
  rank: number;
  arrow_numbers: number[];
  consistency_score: number;
  weight_std_gr: number;
  spine_std: number;
  straightness_std: number;
}

/**
 * Find the top N most consistent sets of set_size arrows.
 * Uses exhaustive search for small N; random sampling when combinations exceed maxCombinations.
 */
export function findBestSets(
  shafts: ShaftData[],
  setSize = 6,
  topN = 5,
  weights?: Partial<typeof DEFAULT_WEIGHTS>,
  maxCombinations = 200_000,
): BestSetResult[] {
  const n = shafts.length;
  if (n < setSize) return [];

  // Generate combinations
  const candidates: number[][] = [];

  function* combinations(arr: number[], r: number): Generator<number[]> {
    if (r === 0) { yield []; return; }
    for (let i = 0; i <= arr.length - r; i++) {
      for (const rest of combinations(arr.slice(i + 1), r - 1)) {
        yield [arr[i], ...rest];
      }
    }
  }

  // Estimate total combinations (n choose setSize)
  function comb(n: number, k: number): number {
    if (k > n) return 0;
    if (k === 0 || k === n) return 1;
    let result = 1;
    for (let i = 0; i < Math.min(k, n - k); i++) {
      result = (result * (n - i)) / (i + 1);
    }
    return Math.round(result);
  }

  const totalCombos = comb(n, setSize);

  if (totalCombos <= maxCombinations) {
    const indices = Array.from({ length: n }, (_, i) => i);
    for (const combo of combinations(indices, setSize)) {
      candidates.push(combo);
    }
  } else {
    // Random sampling
    const seen = new Set<string>();
    const indices = Array.from({ length: n }, (_, i) => i);
    while (seen.size < maxCombinations) {
      const shuffled = [...indices].sort(() => Math.random() - 0.5);
      const combo = shuffled.slice(0, setSize).sort((a, b) => a - b);
      const key = combo.join(",");
      if (!seen.has(key)) {
        seen.add(key);
        candidates.push(combo);
      }
    }
  }

  // Score all candidates
  const scored: Array<[number, number[]]> = candidates.map((combo) => {
    const subset = combo.map((i) => shafts[i]);
    return [scoreSetConsistency(subset, weights), combo];
  });

  scored.sort((a, b) => a[0] - b[0]);

  return scored.slice(0, topN).map(([consistency, combo], idx) => {
    const subset = combo.map((i) => shafts[i]);
    const arrowNumbers = subset.map((s) => s.arrow_number ?? 0).sort((a, b) => a - b);

    const weightVals = subset.filter((s) => s.measured_weight_gr != null).map((s) => s.measured_weight_gr!);
    const spineVals = subset.filter((s) => s.measured_spine_astm != null).map((s) => s.measured_spine_astm!);
    const straightVals = subset.filter((s) => s.straightness != null).map((s) => s.straightness!);

    return {
      rank: idx + 1,
      arrow_numbers: arrowNumbers,
      consistency_score: consistency,
      weight_std_gr: stdPop(weightVals),
      spine_std: stdPop(spineVals),
      straightness_std: stdPop(straightVals),
    };
  });
}

// ---------------------------------------------------------------------------
// Similar arrows finder
// ---------------------------------------------------------------------------

export interface SimilarArrowResult {
  arrow_number: number;
  similarity_score: number;
  weight_diff_gr: number;
  spine_diff: number;
  straightness_diff: number;
}

/**
 * Find arrows most similar to a reference set's centroid.
 */
export function findSimilarArrows(
  referenceArrows: ShaftData[],
  candidates: ShaftData[],
  topN = 5,
): SimilarArrowResult[] {
  if (referenceArrows.length === 0 || candidates.length === 0) return [];

  const features = ["measured_weight_gr", "measured_spine_astm", "straightness"] as const;

  // Compute centroid
  const centroid: Record<string, number> = {};
  for (const feat of features) {
    const vals = referenceArrows.filter((a) => a[feat] != null).map((a) => a[feat] as number);
    centroid[feat] = vals.length > 0 ? vals.reduce((a, b) => a + b, 0) / vals.length : 0.0;
  }

  // Normalization ranges
  const allArrows = [...referenceArrows, ...candidates];
  const ranges: Record<string, [number, number]> = {};
  for (const feat of features) {
    const vals = allArrows.filter((a) => a[feat] != null).map((a) => a[feat] as number);
    if (vals.length > 0) {
      ranges[feat] = [Math.min(...vals), Math.max(...vals)];
    } else {
      ranges[feat] = [0.0, 1.0];
    }
  }

  function normalize(value: number, feat: string): number {
    const [lo, hi] = ranges[feat];
    return hi !== lo ? (value - lo) / (hi - lo) : 0.5;
  }

  const centroidNorm: Record<string, number> = {};
  for (const feat of features) {
    centroidNorm[feat] = normalize(centroid[feat], feat);
  }

  // Rank candidates
  const scored: Array<[number, SimilarArrowResult]> = [];
  for (const arrow of candidates) {
    let skip = false;
    const valsNorm: Record<string, number> = {};
    const rawVals: Record<string, number> = {};

    for (const feat of features) {
      const v = arrow[feat];
      if (v == null) { skip = true; break; }
      valsNorm[feat] = normalize(v as number, feat);
      rawVals[feat] = v as number;
    }
    if (skip) continue;

    const dist = Math.sqrt(features.reduce((acc, f) => acc + (valsNorm[f] - centroidNorm[f]) ** 2, 0));
    const maxDist = Math.sqrt(features.length);
    const similarity = Math.max(0.0, (1.0 - dist / maxDist) * 100.0);

    scored.push([similarity, {
      arrow_number: arrow.arrow_number ?? 0,
      similarity_score: Math.round(similarity * 10) / 10,
      weight_diff_gr: Math.round((rawVals.measured_weight_gr - centroid.measured_weight_gr) * 100) / 100,
      spine_diff: Math.round((rawVals.measured_spine_astm - centroid.measured_spine_astm) * 100) / 100,
      straightness_diff: Math.round((rawVals.straightness - centroid.straightness) * 10000) / 10000,
    }]);
  }

  scored.sort((a, b) => b[0] - a[0]);
  return scored.slice(0, topN).map(([, result]) => result);
}
